JPEG/XMP: Parse GContainer additional images

cjee21 opened the pull request:
https://developer.android.com/media/platform/hdr-image-format#GContainer_details

cjee21 commented:
Not done yet. I have not figured out how to parse the images yet. They have size and padding but not offset from beginning of file and the size for primary image is optional too. Need to also handle case where both GContainer and MPF is referring to the same image as well as GContainer secondary image may contain main MPF directory.

cjee21 commented:
Done. May not be the best code but it works at least with the Google Pixel 9 Pro samples.

JeromeMartinez commented:
>  the secondary "original image" also contains a main MPF

a main MPF in a secondary image, head hache... I check.

cjee21 commented:
Fixed typo for composite image and mapped the values.
Fixed JPEG stream merging. There is case of secondary images having their own Exif thumbnails so there are now shown but not indicated that is is from which stream yet.
Removed some redundant code after the merging of the two MPF IFDs handling.

JeromeMartinez commented on Source/MediaInfo/Tag/File_Exif.cpp:
```diff
@@ -1262,6 +1262,18 @@ static const char* Exif_IFDExif_WhiteBalance_Name(int16u value)
+static const char* Exif_IFDExif_CompositeImage_Name(int16u value)
+{
+    switch (value) {
+    case 0: return "Unknown";
```
remove this line.

cjee21 commented:
I added the parsing for the yet another method and mapped them to similar 'API' as GContainer. However this one is in extended XMP so requires more changes to JPEG side.

JeromeMartinez commented:
> 3D MPFs I collected online do have dependents in MPF and Exif thumbnails in secondary MPFs.

About that... I wonder what are the "Disparity" images?
I am wondering if I shouldn't "merge" some image tracks and show "MultiView" instead, as the idea is to show what the user would see. Maybe in another PR, first we catch all images!

cjee21 commented:
> About that... I wonder what are the "Disparity" images?

For 3D and Panorama, when there are two disparity, it is 3D, one image for each eye. There is additional data in the other MPF directory about convergence etc.

For multiview, the many images are different angles. When viewed on Sony devices (older cameras and Xperia), you can swivel the device left/right and it will look like 3D as it shows the different angles. https://www.youtube.com/watch?v=GLI0u_9M9s0

JeromeMartinez commented:
> I added the parsing for the yet another method and mapped them to similar 'API' as GContainer. However this one is in extended XMP so requires more changes to JPEG side.

Maybe I screw up something but I don't see the other images for Google Pixel 4 and Google Pixel 7a.

cjee21 commented:
Well the 2nd MPF thumbnail is dependent of Image 4 but is not technically 'Muxed in Image 4' right?

JeromeMartinez commented:
> Well the 2nd MPF thumbnail is dependent of Image 4 but is not technically 'Muxed in Image 4' right?

Right. I reused something I do for other topics (for instance text in video stream) as a quick implementation, I'll see later how to better show that.

>  I do not know how to modify the Extended XMP part....

I check if I can do something quickly.

> There is additional data in the other MPF directory about convergence etc.

It is all "1" in MPF metadata of all images of "3D Multiview dsc00191.jpg", or do I miss something?

cjee21 commented:
> It is all "1" in MPF metadata of all images of "3D Multiview dsc00191.jpg", or do I miss something?

Ya I remember seeing all '1'.

MPF specs:

![Screenshot 2025-06-19 180913](https://github.com/user-attachments/assets/593a1fdd-d470-4b52-be9f-023d081ad547)
![Screenshot 2025-06-19 180930](https://github.com/user-attachments/assets/2b793219-3356-43fb-8d3d-4dc72842630b)

JeromeMartinez commented:
> The required changes to JPEG side is not done yet. I just parsed the XMP part. I do not know how to modify the Extended XMP part....

Actually it is same for XMP or Extended XMP, just that there is no MPF in the 4/7a files so we need to find the start of the extra images ourselves, I added a WIP commit for that so you can test on your side.
Known issue is lack of general stream size and conformance error for 6824240430.jpg.

JeromeMartinez commented:
Parsing up to the EOI doesn't work in practice due to lack of complete JPEG parser + I see that the MP4 is before the EOI (???) with the Google Pixel 7a file.
Computing from the EOF as suggested in some Google specs does not work because the Google Pixel 4 & 7a have [aecDebug content](https://learn.microsoft.com/en-us/azure/communication-services/resources/troubleshooting/voice-video-calling/references/how-to-collect-diagnostic-audio-recordings) after the data listed in GContainer

It is not important enough for me for spending more time on that so for the moment I just list the data found in the GContainer, without parsing the target data.

BTW, thank you for the other PRs, very appreciated but I can not check them quickly due to work I have to do on other parts of MI, but keep sending PR if you can, I'll review them at some point.

cjee21 commented:
> Let's try with a separate parser.

Sneak peek:

<details><summary>Details</summary>
<p>

```
406E4B FFD8 - SOI - Start Of Image (2 bytes)
406E4B  Header (2 bytes)
406E4B   Marker:                               65496 (0xFFD8)
406E4D FFE2 - APP2 - Application-specific marker 2 - ISO 21496-1 Gain map metadata for image conversion (173 bytes)
406E4D  Header (4 bytes)
406E4D   Marker:                               65506 (0xFFE2)
406E4F   Fl - Frame header length:             171 (0x00AB)
406E51  Signature:                             urn:iso:std:iso:ts:21496:-1
406E6D  ISO 21496-1 Metadata (141 bytes)
406E6D   min_version:                          0 (0x0000)
406E6F   writer_version:                       0 (0x0000)
406E71   flags:                                192 (0xC0)
406E72   backward_direction:                   No
406E72   is_multi_channel:                     Yes
406E72   use_base_color_space:                 Yes
406E72   use_common_denominator:               No
406E72   base_hdr_headroom_n:                  0 (0x00000000)
406E76   base_hdr_headroom_d:                  1 (0x00000001)
406E7A   alternate_hdr_headroom_n:             83262 (0x0001453E)
406E7E   alternate_hdr_headroom_d:             32768 (0x00008000)
406E82   channel 1 (40 bytes)
406E82    gain_map_min_n:                      4230153492 (0xFC230514)
406E86    gain_map_min_d:                      1073741824 (0x40000000)
406E8A    gain_map_max_n:                      73697 (0x00011FE1)
406E8E    gain_map_max_d:                      32768 (0x00008000)
406E92    gain_map_gamma_n:                    273391370 (0x104B9F0A)
406E96    gain_map_gamma_d:                    1073741824 (0x40000000)
406E9A    base_offset_n:                       16777216 (0x01000000)
406E9E    base_offset_d:                       1073741824 (0x40000000)
406EA2    alternate_offset_n:                  16777216 (0x01000000)
406EA6    alternate_offset_d:                  1073741824 (0x40000000)
406EAA   channel 2 (40 bytes)
406EAA    gain_map_min_n:                      4259014660 (0xFDDB6804)
406EAE    gain_map_min_d:                      1073741824 (0x40000000)
406EB2    gain_map_max_n:                      69992 (0x00011168)
406EB6    gain_map_max_d:                      32768 (0x00008000)
406EBA    gain_map_gamma_n:                    271120723 (0x1028F953)
406EBE    gain_map_gamma_d:                    1073741824 (0x40000000)
406EC2    base_offset_n:                       16777216 (0x01000000)
406EC6    base_offset_d:                       1073741824 (0x40000000)
406ECA    alternate_offset_n:                  16777216 (0x01000000)
406ECE    alternate_offset_d:                  1073741824 (0x40000000)
406ED2   channel 3 (40 bytes)
406ED2    gain_map_min_n:                      4145445776 (0xF7167B90)
406ED6    gain_map_min_d:                      1073741824 (0x40000000)
406EDA    gain_map_max_n:                      69530 (0x00010F9A)
406EDE    gain_map_max_d:                      32768 (0x00008000)
406EE2    gain_map_gamma_n:                    311797823 (0x1295A83F)
406EE6    gain_map_gamma_d:                    1073741824 (0x40000000)
406EEA    base_offset_n:                       16777216 (0x01000000)
406EEE    base_offset_d:                       1073741824 (0x40000000)
406EF2    alternate_offset_n:                  16777216 (0x01000000)
406EF6    alternate_offset_d:                  1073741824 (0x40000000)
406EFA FFDB - DQT - Define Quantization Tables (134 bytes)
...
```

</p>
</details>
